### PR TITLE
feat: variables hovering

### PR DESCRIPTION
**Pull request by grianbrcom**
Variables hovering shows it's structure

### CODE REVIEW

File: server/src/main/java/com/broadcom/lsp/cobol/service/delegates/hover/VariableHover.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      if (RangeUtils.isInside(position, variable.getDefinition().toLocation())
+          || variable.getUsages().stream()
+               .map(Locality::toLocation)
+               .anyMatch(location -> RangeUtils.isInside(position, location)))
+        return variable;
+    return null;
+  }
+
+  private static List<String> parentsDescriptions(Variable variable) {
+    if (variable == null) return new ArrayList<>();
```
Comment (asatklichov) [inline server/src/main/java/com/broadcom/lsp/cobol/service/delegates/hover/VariableHover.java:70]: In below childrenDescriptions returned  Collections.emptyList(), I guess you will modify this later, right?

Comment (asatklichov) [reply server/src/main/java/com/broadcom/lsp/cobol/service/delegates/hover/VariableHover.java:70]: And there are conflicts

File: server/src/main/java/com/broadcom/lsp/cobol/core/model/variables/ElementItem.java [hunk 2]
```diff
@@ -48,11 +49,25 @@ public ElementItem(
...
+
+  @Override
+  public String getDisplayLine() {
+    StringBuilder stringBuilder = new StringBuilder();
+    stringBuilder.append(String.format("%1$02d %2$s", levelNumber, name));
+    if (picClause != null)
+      stringBuilder.append(" PIC ").append(picClause);
+    if (usageFormat != UsageFormat.UNDEFINED)
+      stringBuilder.append(" USAGE ").append(usageFormat);
+    if (value != null && !value.isEmpty())
```
[the remark this answers is not in the document]
Comment (grianbrcom) [reply server/src/main/java/com/broadcom/lsp/cobol/core/model/variables/ElementItem.java:69]: Indeed!

File: server/src/main/java/com/broadcom/lsp/cobol/core/model/variables/MultiTableDataName.java [hunk 3]
```diff
@@ -59,4 +60,13 @@ public Variable rename(RenameItem newParent) {
         indexes,
         usageFormat);
   }
+
+  @Override
+  public String getDisplayLine() {
```
[the remark this answers is not in the document]
Comment (grianbrcom) [reply server/src/main/java/com/broadcom/lsp/cobol/core/model/variables/MultiTableDataName.java:65]: Yes, makes sense.

File: server/src/main/java/com/broadcom/lsp/cobol/service/delegates/hover/VariableHover.java (hunk 1 above)
Comment (grianbrcom) [reply server/src/main/java/com/broadcom/lsp/cobol/service/delegates/hover/VariableHover.java:70]: Yes, I call `parentsDescriptions` recursively and add parents into this array. And `childrenDescriptions` called only once.

File: server/src/main/java/com/broadcom/lsp/cobol/core/model/variables/ElementItem.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (asatklichov) [reply server/src/main/java/com/broadcom/lsp/cobol/core/model/variables/ElementItem.java:69]: > nde

Yes, we have that library already, can be used

File: server/src/main/java/org/eclipse/lsp/cobol/core/model/variables/ConditionDataName.java [hunk 4]
```diff
@@ -56,6 +60,15 @@ public void addConditionName(ConditionDataName variable) {
...
   public Variable rename(RenameItem newParent) {
-    return new ConditionDataName(name, definition, newParent, value);
+    return new ConditionDataName(name, definition, newParent, value, valueTo);
+  }
+
+  @Override
+  public String getFormattedDisplayLine() {
+    StringBuilder stringBuilder = new StringBuilder();
+    stringBuilder.append(String.format("%1$02d %2$s VALUE %3$s", levelNumber, name, value));
+    if (valueTo != null)
```
Comment (ishche) [inline server/src/main/java/org/eclipse/lsp/cobol/core/model/variables/ConditionDataName.java:70]: Are you allow ifs without curly brackets in che4z codebase?

Comment (grianbrcom) [reply server/src/main/java/org/eclipse/lsp/cobol/core/model/variables/ConditionDataName.java:70]: Yes, I clarified it when I introduced the style check system.
And personally, I like it when curly brackets are not mandatory. So I was not arguing a lot)

**Review by asatklichov — APPROVED** [review]
🐎

File: server/src/test/java/org/eclipse/lsp/cobol/service/delegates/references/ElementOccurrencesTest.java [hunk 5]
```diff
@@ -213,7 +214,7 @@ void variousCases(AnalysisResult analysisResult, Position position, List<Locatio
     @ParameterizedTest
     @MethodSource("insideTestData")
     void isInsideTest(String uri, Position position, Location location, boolean result) {
-        assertEquals(result, ElementOccurrences.isInside(
+        assertEquals(result, RangeUtils.isInside(
```
[the remark this answers is not in the document]
Comment (grianbrcom) [reply server/src/test/java/org/eclipse/lsp/cobol/service/delegates/references/ElementOccurrencesTest.java:217]: The whole file was not aligned. I've fixed it.

File: server/src/main/java/org/eclipse/lsp/cobol/core/model/variables/IndexItem.java [hunk 6]
```diff
@@ -28,7 +28,25 @@
 @ToString(callSuper = true)
 @EqualsAndHashCode(callSuper = true)
 public class IndexItem extends AbstractVariable {
-  public IndexItem(String name, Locality definition) {
-    super(name, definition, null);
+  public IndexItem(int levelNumber, String name, Locality definition) {
```
[the remark this answers is not in the document]
Comment (grianbrcom) [reply server/src/main/java/org/eclipse/lsp/cobol/core/model/variables/IndexItem.java:31]: Here is the trick in VariableDefinitionDelegate:
```
  @NonNull
  private List<IndexItem> retrieveIndexItem(int levelNumber, @NonNull DataOccursClauseContext clause) {
    return clause.indexName().stream()
        .map(it -> new IndexItem(levelNumber, it.getText().toUpperCase(), positions.get(it.start)))
        .collect(toList());
  }
```
And we use it like this:
```
      MultiTableDataName result =
          new MultiTableDataName(
               <...>
              retrieveIndexItem(variable.getNumber(), variable.getOccursClauses().get(0)),
```
So, as you see we construct `IndexItem`, but the parent `MultiTableDataName` is not constructed yet. Both classes are unmodifable. That's why I decided to create `IndexItem` without parent first and then recreate with the correct parent inside the `MultiTableDataName` constructor:
```
  public MultiTableDataName(
      int levelNumber,
      String name,
      Locality definition,
      Variable parent,
      int occursTimes,
      List<IndexItem> indexes,
      UsageFormat usageFormat) {
    super(levelNumber, name, definition, parent);
    this.occursTimes = occursTimes;
    this.indexes = indexes.stream().map(it -> it.updateParent(this)).collect(Collectors.toList());
    this.usageFormat = usageFormat;
  }
```
And I've forgotten about `TableDataName`, will fix it.

File: server/src/main/java/org/eclipse/lsp/cobol/core/model/variables/MnemonicName.java [hunk 7]
```diff
@@ -32,8 +32,8 @@
 @ToString(callSuper = true)
 @EqualsAndHashCode(callSuper = true)
 public class MnemonicName extends AbstractVariable {
-  public MnemonicName(String name, Locality definition) {
-    super(name, definition, null);
+  public MnemonicName(int levelNumber, String name, Locality definition) {
+    super(levelNumber, name, definition, null);
```
[the remark this answers is not in the document]
Comment (grianbrcom) [reply server/src/main/java/org/eclipse/lsp/cobol/core/model/variables/MnemonicName.java:36]: It was discussed personally.

File: server/src/main/java/org/eclipse/lsp/cobol/core/semantics/outline/RangeUtils.java [hunk 8]
```diff
@@ -91,4 +89,16 @@ public boolean isAfter(Position before, Position after) {
...
+
+  /**
+   * Tests if text document position located inside the location.
+   * @param position - the document position
+   * @param location - en element location
+   * @return true if position is inside the location.
+   */
+  public boolean isInside(TextDocumentPositionParams position, Location location) {
+    return position.getTextDocument().getUri().equals(location.getUri())
+        && !RangeUtils.isBefore(position.getPosition(), location.getRange().getStart())
```
[the remark this answers is not in the document]
Comment (grianbrcom) [reply server/src/main/java/org/eclipse/lsp/cobol/core/semantics/outline/RangeUtils.java:101]: Ooops)
Because of copy/paste...

**Review by temanbrcom — APPROVED** [review]
☄️

**Pull request merged**